src/integer_mod_q/mat_zq/arithmetic/add.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright © 2023 Phil Milewski
+//
+// This file is part of qFALL-math.
+//
+// qFALL-math is free software: you can redistribute it and/or modify it under
+// the terms of the Mozilla Public License Version 2.0 as published by the
+// Mozilla Foundation. See <https://mozilla.org/en-US/MPL/2.0/>.
+
+//! Implementation of the [`Add`] trait for [`MatZq`] values.
+
+use super::super::MatZq;
+use crate::error::MathError;
+use crate::macros::arithmetics::{
+    arithmetic_trait_borrowed_to_owned, arithmetic_trait_mixed_borrowed_owned,
+};
+use crate::traits::{GetNumColumns, GetNumRows};
+use flint_sys::fmpz_mod_mat::fmpz_mod_mat_add;
+use std::ops::Add;
+
+impl Add for &MatZq {
+    type Output = MatZq;
+    /// Implements the [`Add`] trait for two [`MatZq`] values.
+    /// [`Add`] is implemented for any combination of [`MatZq`] and borrowed [`MatZq`].
+    ///
+    /// Parameters:
+    /// - `other`: specifies the value to add to `self`
+    ///
+    /// Returns the sum of both numbers as a [`MatZq`].
+    ///
+    /// # Example
+    /// ```
+    /// use qfall_math::integer_mod_q::MatZq;
+    /// use std::str::FromStr;
+    ///
+    /// let a: MatZq = MatZq::from_str("[[1, 2, 3],[3, 4, 5]] mod 7").unwrap();
+    /// let b: MatZq = MatZq::from_str("[[1, 9, 3],[1, 0, 5]] mod 7").unwrap();
+    ///
+    /// let c: MatZq = &a + &b;
+    /// let d: MatZq = a + b;
+    /// let e: MatZq = &c + d;
+    /// let f: MatZq = c + &e;
+    /// ```
+    ///
+    /// # Panics
+    /// - Panics if the dimensions of both matrices mismatch
+    /// - Panics if the moduli mismatch
+    fn add(self, other: Self) -> Self::Output {
+        self.add_safe(other).unwrap()
+    }
+}
+
+impl MatZq {
+    /// Implements addition for two [`MatZq`] matrixes.
+    ///
+    /// Parameters:
+    /// - `other`: specifies the value to add to `self`
+    ///
+    /// Returns the sum of both matrixes as a [`MatZq`] or an
+    /// error if the matrix dimensions do mismatch.
+    ///
+    /// # Example
+    /// ```
+    /// use qfall_math::integer_mod_q::MatZq;
+    /// use std::str::FromStr;
+    ///
+    /// let a: MatZq = MatZq::from_str("[[1, 2, 3],[3, 4, 5]] mod 7").unwrap();
+    /// let b: MatZq = MatZq::from_str("[[1, 9, 3],[1, 0, 5]] mod 7").unwrap();
+    ///
+    /// let c: MatZq = a.add_safe(&b).unwrap();
+    /// ```
+    /// # Errors
+    /// Returns a [`MathError`] of type
+    /// [`MathError::MismatchingMatrixDimension`] if the matrix dimensions
+    /// mismatch.
+    /// Returns a [`MathError`] of type
+    /// [`MathError::MismatchingModulus`] if the moduli mismatch.
+    pub fn add_safe(&self, other: &Self) -> Result<MatZq, MathError> {
+        if self.get_mod() != other.get_mod() {
+            return Err(MathError::MismatchingModulus(format!(
+                " Tried to add matrixes with moduli '{}' and '{}'.
+            If the modulus should be ignored please convert into a Z beforehand.",
```

Comment:
this comment does not make sense here, please adapt to something more meaningful